windows/src/engine/keyman32/preservedkeymap.cpp

```diff
@@ -318,13 +371,114 @@ BOOL PreservedKeyMap::MapKeyboard(KEYBOARD *pKeyboard, PreservedKey **pPreserved
   return TRUE;
 }
 
+BOOL
+PreservedKeyMap::MapKeyboardCore(km_kbp_keyboard *pKeyboard, PreservedKey **pPreservedKeys, size_t *cPreservedKeys) {
+  size_t cKeys = 0, cRules = 0, n = 0;
+  DWORD i;
+
+  m_BaseKeyboardUsesAltGr = KeyboardGivesCtrlRAltForRAlt();  // I4592
+
+  // This is not the same as m_BaseKeyboardUsesAltGr -- we are turning
+  // the simulation back on after (possibly) turning it off, for a
+  // consistent experience. TODO: determine if m_BaseKeyboardUseAltGr
+  // is still needed given we always use kbdus as base for Keyman 10+
+  BOOL bSimulateAltGr = Globals::get_SimulateAltGr();
+
+  // We only want to translate RALT and RALT+SHIFT for Ctrl+Alt rules.
+  // So we exclude all our other favourite modifier keys.
+  const UINT RALT_MATCHING_MASK = TF_MOD_CONTROL | TF_MOD_ALT | TF_MOD_LCONTROL | TF_MOD_RCONTROL | TF_MOD_LALT | TF_MOD_RALT;
+
+  // This is where we will call down to the api to get list of keys used in the keyboard rules
+  km_kbp_keyboard_key *kb_key_list;
+
+  km_kbp_status err_status = km_kbp_keyboard_get_key_list(pKeyboard, &kb_key_list);
+  if ((err_status != KM_KBP_STATUS_OK) || (kb_key_list ==nullptr)) {
+    return FALSE;
+  }
+
+  km_kbp_keyboard_key *key_rule_it = kb_key_list;
+  for (; key_rule_it->key; ++key_rule_it) {
+    ++cRules;
+  }
+  cKeys = cRules;
+  if (cKeys == 0) {
```

Comment:
Shouldn't we set `*cPreservedKeys = cKeys` here as well?

Reply:
I guess we could check if `pPreservedKeys == NULL` and set `*cPreservedKeys = cKeys` basically setting it to zero. but in this case we are returning FALSE anyway.  So if the function is returning FALSE for failure, I don't think the `*cPreservedKeys` should be trusted.
This also mimics the behaviour of the current `::MapCore`